File: src/api_log_test_mcp/tools/orchestrate.py
"""run_suite: execute a full suite end-to-end and emit a report.

Batched flow (so a rate-limited remote log backend is downloaded only once per run):
  1. For every case: build URL, [attach auth], call_api (stamps a correlation id),
     assert the response.
  2. If any case has ``validate_logs=Yes``: wait for log propagation, then take ONE snapshot
     per distinct ``log_source``, validate each such case's ``expected_log_strings`` against it
     (filtered by that case's correlation id), and merge the result in.
  3. Discard snapshots.

Flags gate the credentialed work: ``auth_required=no`` -> no OAuth token; ``validate_logs=no``
-> no log download/validation. Each case is isolated so one failure never aborts the run.
"""

from __future__ import annotations

import shutil
import time
import uuid
from dataclasses import dataclass
from pathlib import Path

from ..config import Settings, get_settings
from ..models import ApiResponse, CaseEvidence, CaseReport, SuiteReport, TestCase
from .auth import get_auth_token
from .http_runner import ApiCallError, call_api
from .logs import (
    correlation_present,
    discard_snapshot,
    matched_log_lines,
    snapshot_logs,
    validate_logs,
)
from .response import assert_case_response
from .results_writer import write_evidence_tabs, write_results
from .suite import read_test_suite


@dataclass
class _CaseRun:
    """A case's request-phase outcome, carried into the log-validation phase."""

    case: TestCase
    report: CaseReport
    correlation_id: str | None
    sent_request: dict | None = None
    response: ApiResponse | None = None
    matched_log_lines: dict[str, list[str]] | None = None


def run_suite(suite_path: str, retain_snapshots: bool = False) -> SuiteReport:
    """Run every case in the suite at ``suite_path`` and return an aggregate report."""
    report, _evidence = _run(suite_path, retain_snapshots)
    return report


def run_and_record(
    suite_path: str, retain_snapshots: bool = False, results_path: str | None = None
) -> tuple[SuiteReport, str, str]:
    """Run the suite and record results into a SEPARATE ``<stem>_results.xlsx`` file.

    Use this (rather than bare ``run_suite``) so every test run is recorded. The **suite file is
    never modified** — results are written to a sibling results workbook (``results_path`` if
    given, else ``_results_path(suite_path)``). The results file is cloned from the suite on first
    use (so it carries the case definitions), then each run appends a timestamped ``RESULTS``
    summary block and overwrites the per-case evidence tabs (latest run only). Returns
    ``(report, run_at, results_path)``.
    """
    report, evidence = _run(suite_path, retain_snapshots)
    out = Path(results_path) if results_path else _results_path(suite_path)
    if not out.exists():
        shutil.copy2(suite_path, out)  # seed the results file with the suite's case definitions
    run_at = write_results(str(out), report)
    write_evidence_tabs(str(out), evidence, run_at)
    return report, run_at, str(out)


def _results_path(suite_path: str) -> Path:
    """Sibling results workbook for ``suite_path``: ``<stem>_results<suffix>``.

    A trailing ``_suite`` in the stem is replaced (so ``foo_suite.xlsx`` -> ``foo_results.xlsx``);
    otherwise ``_results`` is appended (``foo.xlsx`` -> ``foo_results.xlsx``).
    """
    p = Path(suite_path)
    stem = p.stem[: -len("_suite")] if p.stem.endswith("_suite") else p.stem
    return p.with_name(f"{stem}_results{p.suffix}")


def _run(suite_path: str, retain_snapshots: bool = False) -> tuple[SuiteReport, list[CaseEvidence]]:
    """Run every case and return both the aggregate report and per-case evidence."""
    import os
    os.environ["_CURRENT_SUITE_PATH"] = suite_path

    suite = read_test_suite(suite_path)
    settings = get_settings()
    report = SuiteReport(parse_errors=suite.parse_errors)

    # Phase 1: requests + response assertions.
    runs = [_run_request(case, suite.base_path) for case in suite.cases]

    # Phase 2: log validation (one snapshot per distinct source, reused across cases).
    active_snapshots: list[str] = []
    try:
        _validate_logs_phase(runs, settings, active_snapshots, suite.application_logs_fetch_url)
    finally:
        if not retain_snapshots:
            for sid in active_snapshots:
                discard_snapshot(sid)

    report.cases = [r.report for r in runs]
    report.total = len(report.cases)
    report.passed = sum(1 for c in report.cases if c.passed)
    report.failed = report.total - report.passed
    return report, [_build_evidence(r) for r in runs]


def _build_evidence(run: _CaseRun) -> CaseEvidence:
    """Flatten a finished ``_CaseRun`` into the self-contained evidence record for its tab."""
    case, rep = run.case, run.report
    ra, lv, resp = rep.response_assert, rep.log_validation, run.response
    req = run.sent_request or {}
    return CaseEvidence(
        test_id=case.test_id,
        description=case.description,
        passed=rep.passed,
        error=rep.error,
        method=req.get("method"),
        url=req.get("url"),
        request_headers=req.get("headers") or {},
        request_body=req.get("body"),
        actual_status=rep.actual_status,
        expected_status=rep.expected_status,
        latency_ms=resp.latency_ms if resp else None,
        match_mode=ra.mode if ra else None,
        response_passed=ra.passed if ra else None,
        response_diffs=ra.diffs if ra else [],
        expected_response=case.expected_response,
        actual_body=resp.body if resp else None,
        validated_logs=case.validate_logs,
        logs_passed=lv.passed if lv else None,
        log_source=case.log_source if case.validate_logs else None,
        correlation_id=rep.correlation_id,
        expected_log_strings=case.expected_log_strings,
        matched_logs=lv.matched if lv else [],
        missing_logs=lv.missing if lv else [],
        used_fallback=lv.used_fallback if lv else False,
        lines_considered=lv.lines_considered if lv else 0,
        matched_log_lines=run.matched_log_lines or {},
    )


def _join_url(base_path: str | None, url: str) -> str:
    """Join a suite base path with a case url, preserving the base subpath.

    ``urljoin`` discards the base subpath when ``url`` starts with ``/`` (it treats it as
    host-absolute), so ``.../api`` + ``/products`` would wrongly become ``.../products``.
    A fully-qualified case url (``http(s)://...``) is used as-is.
    """
    if url.startswith(("http://", "https://")) or not base_path:
        return url
    return f"{base_path.rstrip('/')}/{url.lstrip('/')}"


def _run_request(case: TestCase, base_path: str | None) -> _CaseRun:
    """Do the call + response assertion for one case (no log validation yet)."""
    # Generate the correlation id up front so it is recorded even if the request fails.
    correlation_id = f"{case.test_id}-{uuid.uuid4().hex[:12]}"
    try:
        url = _join_url(base_path, case.url)

        headers = dict(case.headers)
        if case.auth_required:
            headers["Authorization"] = f"Bearer {get_auth_token()}"

        sent_request = {"method": case.method, "url": url, "headers": headers, "body": case.body}
        response = call_api(
            case.method, url, headers=headers, body=case.body, correlation_id=correlation_id,
        )
        response_assert = assert_case_response(case, response)
        report = CaseReport(
            test_id=case.test_id,
            passed=response_assert.passed,
            correlation_id=correlation_id,
            actual_status=response.status,
            expected_status=case.expected_status,
            response_assert=response_assert,
        )
        return _CaseRun(case=case, report=report, correlation_id=correlation_id,
                        sent_request=sent_request, response=response)
    except ApiCallError as exc:
        return _failed_run(case, correlation_id, f"request failed: {exc}")
    except NotImplementedError as exc:
        return _failed_run(case, correlation_id, str(exc))
    except Exception as exc:  # noqa: BLE001 - any case error must not abort the suite
        return _failed_run(case, correlation_id, f"{type(exc).__name__}: {exc}")


def _failed_run(case: TestCase, correlation_id: str, error: str) -> _CaseRun:
    report = CaseReport(test_id=case.test_id, passed=False, correlation_id=correlation_id,
                        expected_status=case.expected_status, error=error)
    return _CaseRun(case=case, report=report, correlation_id=correlation_id)


def _validate_logs_phase(
    runs: list[_CaseRun],
    settings: Settings,
    active_snapshots: list[str],
    application_logs_fetch_url: str | None,
) -> None:
    """Snapshot once per distinct log_source and validate each opted-in case against it."""
    log_runs = [r for r in runs if r.case.validate_logs and r.correlation_id]
    if not log_runs:
        return

    # CloudHub needs time to surface a request's logs, so wait before the first fetch.
    if settings.propagation_wait_seconds > 0:
        time.sleep(settings.propagation_wait_seconds)

    # Group by source so each backend is downloaded only once.
    by_source: dict[str, list[_CaseRun]] = {}
    for r in log_runs:
        by_source.setdefault(r.case.log_source, []).append(r)

    for source_name, group in by_source.items():
        correlation_ids = [r.correlation_id for r in group if r.correlation_id]
        try:
            sid = _snapshot_with_retry(
                settings, source_name, correlation_ids, application_logs_fetch_url
            )
            active_snapshots.append(sid)
        except Exception as exc:  # noqa: BLE001 - attribute the failure to every case in group
            for r in group:
                r.report.error = (r.report.error or "") + f" log snapshot failed: {exc}"
                r.report.passed = False
            continue

        for r in group:
            # A log line counts only if it carries this case's correlation id AND the expected
            # message — never a whole-log message-only match against unrelated runs. So both
            # validation and evidence are strictly correlation-scoped (no fallback).
            lv = validate_logs(
                sid, r.correlation_id, r.case.expected_log_strings, r.case.log_match_mode,
                correlation_fallback=False,
            )
            r.report.log_validation = lv
            if not lv.passed:
                r.report.passed = False
            # Capture the actual matching lines as evidence (blank if none carry the corr id).
            r.matched_log_lines = matched_log_lines(
                sid, r.correlation_id, r.case.expected_log_strings, r.case.log_match_mode,
                correlation_fallback=False,
            )


def _snapshot_with_retry(
    settings: Settings,
    source_name: str,
    correlation_ids: list[str],
    application_logs_fetch_url: str | None,
) -> str:
    """Download a log snapshot, retrying until every correlation id's logs have surfaced.

    CloudHub publishes a request's logs with a lag, so after the initial fetch we re-download
    up to ``log_fetch_max_retries`` more times (waiting ``log_fetch_retry_wait_seconds`` between
    tries) until each case's correlation id appears. Intermediate snapshots are discarded. The
    most recent snapshot id is returned even if some ids never showed up (those cases then fail
    honestly rather than blocking forever).
    """
    sid = snapshot_logs(
        settings, log_source=source_name,
        application_logs_fetch_url=application_logs_fetch_url,
    )
    for _ in range(max(0, settings.log_fetch_max_retries)):
        if all(correlation_present(sid, cid) for cid in correlation_ids):
            break
        if settings.log_fetch_retry_wait_seconds > 0:
            time.sleep(settings.log_fetch_retry_wait_seconds)
        discard_snapshot(sid)
        sid = snapshot_logs(
        settings, log_source=source_name,
        application_logs_fetch_url=application_logs_fetch_url,
    )
    return sid

File: worker/src/orchestrate.ts
/**
 * Orchestration core — the testable, DO-free port of tools/orchestrate.py.
 *
 * The Python orchestrator is one synchronous function with time.sleep() for the propagation
 * wait and the log-fetch retry loop. Here those waits are owned by the JobRunner Durable Object
 * (via this.schedule / DO alarms), so the logic is split into pure pieces the DO composes:
 *
 *   1. runRequestsPhase  — Phase 1: per-case request + response assertion (no logs).
 *   2. distinctSources   — group log-validating cases by log_source (download once each).
 *   3. decideRetry       — the _snapshot_with_retry stop/continue decision, per firing.
 *   4. validateGroup     — strict, correlation-scoped log validation + evidence merge.
 *   5. finalizeReport / buildEvidence — aggregate.
 *
 * All network/IO is injected so this module is fully unit-testable offline, exactly like the
 * Python tests that monkeypatch call_api / snapshot_logs / time.sleep.
 */
import { assertResponse } from "./matching/response-matcher.js";
import { callApi as defaultCallApi, ApiCallError } from "./http/runner.js";
import type { FetchLike } from "./http/runner.js";
import {
  matchedLogLines,
  type Snapshot,
  validateLogs,
} from "./logs/snapshot.js";
import type {
  ApiResponse,
  CaseEvidence,
  CaseReport,
  ParseError,
  SuiteReport,
  TestCase,
  TestSuite,
} from "./models.js";

/** The auth stub — Phase 3 in the Python project; still unimplemented. */
export class NotImplementedError extends Error {}
export function getAuthTokenStub(): never {
  throw new NotImplementedError(
    "get_auth_token is a Phase 3 stub (OAuth2 client-credentials); blocked on Gate-0 OAuth details.",
  );
}

export interface CaseRun {
  case: TestCase;
  report: CaseReport;
  correlation_id: string | null;
  sent_request: { method: string; url: string; headers: Record<string, string>; body: unknown } | null;
  response: ApiResponse | null;
  matched_log_lines: Record<string, string[]> | null;
}

export interface RequestDeps {
  callApi: (method: string, url: string, opts: { headers: Record<string, string>; body: unknown; correlationId: string }) => Promise<ApiResponse>;
  getAuthToken: () => Promise<string> | string;
  newCorrelationId: (testId: string) => string;
}

export function defaultRequestDeps(fetchFn?: FetchLike): RequestDeps {
  return {
    callApi: (method, url, opts) => defaultCallApi(method, url, { ...opts, fetchFn }),
    getAuthToken: getAuthTokenStub,
    newCorrelationId: (testId) => `${testId}-${randomHex(12)}`,
  };
}

// --- Phase 1: requests + assertions ----------------------------------------------------

export async function runRequestsPhase(suite: TestSuite, deps: RequestDeps): Promise<CaseRun[]> {
  const runs: CaseRun[] = [];
  for (const c of suite.cases) {
    runs.push(await runRequest(c, suite.base_path, deps));
  }
  return runs;
}

export async function runRequest(c: TestCase, basePath: string | null, deps: RequestDeps): Promise<CaseRun> {
  // Generate the correlation id up front so it is recorded even if the request fails.
  const correlationId = deps.newCorrelationId(c.test_id);
  try {
    const url = joinUrl(basePath, c.url);
    const headers: Record<string, string> = {};
    for (const [k, v] of Object.entries(c.headers)) headers[k] = String(v);
    if (c.auth_required) headers.Authorization = `Bearer ${await deps.getAuthToken()}`;

    const sentRequest = { method: c.method, url, headers, body: c.body };
    const response = await deps.callApi(c.method, url, { headers, body: c.body, correlationId });
    const responseAssert = assertResponse({
      actual_body: response.body,
      expected: c.expected_response,
      mode: c.response_match_mode,
      ignore_paths: c.ignore_paths,
      actual_status: response.status,
      expected_status: c.expected_status,
    });
    const report: CaseReport = {
      test_id: c.test_id,
      passed: responseAssert.passed,
      correlation_id: correlationId,
      actual_status: response.status,
      expected_status: c.expected_status,
      response_assert: responseAssert,
      log_validation: null,
      error: null,
    };
    return { case: c, report, correlation_id: correlationId, sent_request: sentRequest, response, matched_log_lines: null };
  } catch (exc) {
    let message: string;
    if (exc instanceof ApiCallError) message = `request failed: ${exc.message}`;
    else if (exc instanceof NotImplementedError) message = exc.message;
    else message = `${exc instanceof Error ? exc.name : "Error"}: ${exc instanceof Error ? exc.message : String(exc)}`;
    return failedRun(c, correlationId, message);
  }
}

function failedRun(c: TestCase, correlationId: string, error: string): CaseRun {
  return {
    case: c,
    report: {
      test_id: c.test_id,
      passed: false,
      correlation_id: correlationId,
      actual_status: null,
      expected_status: c.expected_status,
      response_assert: null,
      log_validation: null,
      error,
    },
    correlation_id: correlationId,
    sent_request: null,
    response: null,
    matched_log_lines: null,
  };
}

export function joinUrl(basePath: string | null, url: string): string {
  if (url.startsWith("http://") || url.startsWith("https://") || !basePath) return url;
  return `${basePath.replace(/\/+$/, "")}/${url.replace(/^\/+/, "")}`;
}

// --- Phase 2: log validation pieces ----------------------------------------------------

/** Cases that opted into log validation and have a correlation id, grouped by source (first-seen order). */
export function distinctSources(runs: CaseRun[]): { source: string; runs: CaseRun[] }[] {
  const order: string[] = [];
  const bySource = new Map<string, CaseRun[]>();
  for (const r of runs) {
    if (!(r.case.validate_logs && r.correlation_id)) continue;
    const src = r.case.log_source;
    if (!bySource.has(src)) {
      bySource.set(src, []);
      order.push(src);
    }
    bySource.get(src)!.push(r);
  }
  return order.map((source) => ({ source, runs: bySource.get(source)! }));
}

/**
 * The _snapshot_with_retry decision, evaluated once per logPhase firing: validate now if every
 * correlation id has surfaced OR the retry budget is exhausted; otherwise retry after the wait.
 */
export function decideRetry(allPresent: boolean, attempt: number, maxRetries: number): "validate" | "retry" {
  return allPresent || attempt >= maxRetries ? "validate" : "retry";
}

/**
 * Correlation-scoped validation of one source's group against a fresh snapshot. When
 * `correlationFallback` is set and no lines carry a case's correlation id, the whole snapshot
 * is searched instead (Python parity: log_correlation_fallback defaults to true there).
 */
export function validateGroup(snap: Snapshot, group: CaseRun[], correlationFallback = false): void {
  for (const r of group) {
    const corr = r.correlation_id!;
    const lv = validateLogs(snap, corr, r.case.expected_log_strings, r.case.log_match_mode, correlationFallback);
    r.report.log_validation = lv;
    if (!lv.passed) r.report.passed = false;
    r.matched_log_lines = matchedLogLines(snap, corr, r.case.expected_log_strings, r.case.log_match_mode, correlationFallback);
  }
}

/** Attribute a snapshot-download failure to every case in the group (matches the Python path). */
export function failGroupSnapshot(group: CaseRun[], errorMessage: string): void {
  for (const r of group) {
    r.report.error = (r.report.error ?? "") + ` log snapshot failed: ${errorMessage}`;
    r.report.passed = false;
  }
}

// --- aggregation -----------------------------------------------------------------------

export function finalizeReport(runs: CaseRun[], parseErrors: ParseError[]): SuiteReport {
  const cases = runs.map((r) => r.report);
  const passed = cases.filter((c) => c.passed).length;
  return { total: cases.length, passed, failed: cases.length - passed, cases, parse_errors: parseErrors };
}

export function buildEvidence(run: CaseRun): CaseEvidence {
  const { case: c, report: rep } = run;
  const ra = rep.response_assert;
  const lv = rep.log_validation;
  const resp = run.response;
  const req = run.sent_request ?? ({} as NonNullable<CaseRun["sent_request"]>);
  return {
    test_id: c.test_id,
    description: c.description,
    passed: rep.passed,
    error: rep.error,
    method: req.method ?? null,
    url: req.url ?? null,
    request_headers: req.headers ?? {},
    request_body: req.body ?? null,
    actual_status: rep.actual_status,
    expected_status: rep.expected_status,
    latency_ms: resp ? resp.latency_ms : null,
    match_mode: ra ? ra.mode : null,
    response_passed: ra ? ra.passed : null,
    response_diffs: ra ? ra.diffs : [],
    expected_response: c.expected_response,
    actual_body: resp ? resp.body : null,
    validated_logs: c.validate_logs,
    logs_passed: lv ? lv.passed : null,
    log_source: c.validate_logs ? c.log_source : null,
    correlation_id: rep.correlation_id,
    expected_log_strings: c.expected_log_strings,
    matched_logs: lv ? lv.matched : [],
    missing_logs: lv ? lv.missing : [],
    used_fallback: lv ? lv.used_fallback : false,
    lines_considered: lv ? lv.lines_considered : 0,
    matched_log_lines: run.matched_log_lines ?? {},
  };
}

function randomHex(n: number): string {
  return crypto.randomUUID().replace(/-/g, "").slice(0, n);
}
